File: packages/json/llm-json-stream-ts/packages/llm-json-stream/src/classes/property_stream_controller.ts
/**
 * Property stream controllers manage the internal state and emission logic
 * for property streams. These are internal classes not exposed in the public API.
 *
 * Uses async iterators as the primary streaming mechanism.
 */

import { JsonStreamParserController } from "./json_stream_parser.js";
import {
    ArrayPropertyStream,
    BooleanPropertyStream,
    NullPropertyStream,
    NumberPropertyStream,
    ObjectPropertyStream,
    PropertyStream,
    StringPropertyStream,
} from "./property_stream.js";

/**
 * Base class for all property stream controllers.
 */
export abstract class PropertyStreamController<T> {
    abstract readonly propertyStream: PropertyStream<T>;
    protected _isClosed = false;
    protected _finalValue: T | undefined = undefined;
    protected _hasFinalValue = false;
    protected completer: {
        promise: Promise<T>;
        resolve: (value: T) => void;
        reject: (error: Error) => void;
    };

    constructor(
        protected parserController: JsonStreamParserController,
        protected propertyPath: string,
    ) {
        // Create a deferred promise
        let resolve!: (value: T) => void;
        let reject!: (error: Error) => void;
        const promise = new Promise<T>((res, rej) => {
            resolve = res;
            reject = rej;
        });
        this.completer = { promise, resolve, reject };
    }

    get isClosed(): boolean {
        return this._isClosed;
    }

    /**
     * Returns true if this controller has completed with a value.
     */
    get hasFinalValue(): boolean {
        return this._hasFinalValue;
    }

    /**
     * Gets the final value if available.
     * Only valid when hasFinalValue is true.
     */
    get finalValue(): T | undefined {
        return this._finalValue;
    }

    onClose(): void {
        this._isClosed = true;
    }

    complete(value: T): void {
        if (!this._isClosed) {
            this._finalValue = value;
            this._hasFinalValue = true;
            this.propertyStream._complete();
            this.completer.resolve(value);
            this._isClosed = true;
        }
    }

    completeError(error: Error): void {
        if (!this._isClosed) {
            this.propertyStream._error(error);
            this.completer.reject(error);
            this._isClosed = true;
        }
    }
}

/**
 * Controller for string property streams.
 * Manages buffering and streaming of string chunks.
 */
export class StringPropertyStreamController
    extends PropertyStreamController<string> {
    readonly propertyStream: StringPropertyStream;
    private buffer = "";

    constructor(
        parserController: JsonStreamParserController,
        propertyPath: string,
    ) {
        super(parserController, propertyPath);
        this.propertyStream = new StringPropertyStream(
            this.completer.promise,
            parserController,
        );
    }

    /**
     * Adds a string chunk to the stream.
     * The chunk is emitted through the async iterator and accumulated in the buffer.
     */
    addChunk(chunk: string): void {
        if (this._isClosed) return;
        this.buffer += chunk;
        this.propertyStream._pushValue(chunk);
    }

    override complete(value?: string): void {
        if (!this._isClosed) {
            // Always use buffer for strings - the delegate sends chunks to the buffer
            // and completes with empty string when done
            this._finalValue = this.buffer;
            this._hasFinalValue = true;
            this.propertyStream._complete();
            this.completer.resolve(this.buffer);
            this._isClosed = true;
        }
    }
}

/**
 * Controller for number property streams.
 * Numbers emit once when the complete value is parsed.
 */
export class NumberPropertyStreamController
    extends PropertyStreamController<number> {
    readonly propertyStream: NumberPropertyStream;

    constructor(
        parserController: JsonStreamParserController,
        propertyPath: string,
    ) {
        super(parserController, propertyPath);
        this.propertyStream = new NumberPropertyStream(
            this.completer.promise,
            parserController,
        );
    }

    override complete(value: number): void {
        if (!this._isClosed) {
            // Emit the value once through the async iterator
            this._finalValue = value;
            this._hasFinalValue = true;
            this.propertyStream._pushValue(value);
            this.propertyStream._complete();
            this.completer.resolve(value);
            this._isClosed = true;
        }
    }
}

/**
 * Controller for boolean property streams.
 * Booleans emit once when the complete value is parsed.
 */
export class BooleanPropertyStreamController
    extends PropertyStreamController<boolean> {
    readonly propertyStream: BooleanPropertyStream;

    constructor(
        parserController: JsonStreamParserController,
        propertyPath: string,
    ) {
        super(parserController, propertyPath);
        this.propertyStream = new BooleanPropertyStream(
            this.completer.promise,
            parserController,
        );
    }

    override complete(value: boolean): void {
        if (!this._isClosed) {
            // Emit the value once through the async iterator
            this._finalValue = value;
            this._hasFinalValue = true;
            this.propertyStream._pushValue(value);
            this.propertyStream._complete();
            this.completer.resolve(value);
            this._isClosed = true;
        }
    }
}

/**
 * Controller for null property streams.
 * Null emits once when parsed.
 */
export class NullPropertyStreamController
    extends PropertyStreamController<null> {
    readonly propertyStream: NullPropertyStream;

    constructor(
        parserController: JsonStreamParserController,
        propertyPath: string,
    ) {
        super(parserController, propertyPath);
        this.propertyStream = new NullPropertyStream(
            this.completer.promise,
            parserController,
        );
    }

    override complete(value: null): void {
        if (!this._isClosed) {
            // Emit the value once through the async iterator
            this._finalValue = value;
            this._hasFinalValue = true;
            this.propertyStream._pushValue(value);
            this.propertyStream._complete();
            this.completer.resolve(value);
            this._isClosed = true;
        }
    }
}

/**
 * Controller for object property streams.
 * Objects emit snapshots as properties complete.
 */
export class ObjectPropertyStreamController
    extends PropertyStreamController<Record<string, any>> {
    readonly propertyStream: ObjectPropertyStream;
    private currentValue: Record<string, any> = {};

    constructor(
        parserController: JsonStreamParserController,
        propertyPath: string,
    ) {
        super(parserController, propertyPath);
        this.propertyStream = new ObjectPropertyStream(
            this.completer.promise,
            parserController,
            propertyPath,
        );
    }

    /**
     * Notifies that a new property has started parsing.
     */
    notifyProperty(property: PropertyStream<any>, key: string): void {
        this.propertyStream._notifyProperty(property, key);
    }

    /**
     * Adds a property value to the map and emits a snapshot.
     */
    addProperty(key: string, value: any): void {
        if (this._isClosed) return;
        this.currentValue[key] = value;
        // Emit snapshot through async iterator
        this.propertyStream._pushValue({ ...this.currentValue });
    }

    override complete(value?: Record<string, any>): void {
        if (!this._isClosed) {
            const finalValue = value !== undefined ? value : this.currentValue;
            this._finalValue = finalValue;
            this._hasFinalValue = true;
            // Emit the final value through the async iterator before completing
            this.propertyStream._pushValue({ ...finalValue });
            this.propertyStream._complete();
            this.completer.resolve(finalValue);
            this._isClosed = true;
        }
    }
}

/**
 * Controller for array property streams.
 * Arrays emit snapshots as elements complete.
 */
export class ArrayPropertyStreamController<T = any>
    extends PropertyStreamController<T[]> {
    readonly propertyStream: ArrayPropertyStream<T>;
    private currentValue: T[] = [];

    constructor(
        parserController: JsonStreamParserController,
        propertyPath: string,
    ) {
        super(parserController, propertyPath);
        this.propertyStream = new ArrayPropertyStream<T>(
            this.completer.promise,
            parserController,
            propertyPath,
        );
    }

    /**
     * Notifies that a new element has started parsing.
     */
    notifyElement(propertyStream: PropertyStream<any>, index: number): void {
        this.propertyStream._notifyElement(propertyStream, index);
    }

    /**
     * Adds an element to the list and emits a snapshot.
     */
    addElement(value: T): void {
        if (this._isClosed) return;
        this.currentValue.push(value);
        // Emit snapshot through async iterator
        this.propertyStream._pushValue([...this.currentValue]);
    }

    override complete(value?: T[]): void {
        if (!this._isClosed) {
            const finalValue = value !== undefined ? value : this.currentValue;
            this._finalValue = finalValue;
            this._hasFinalValue = true;
            // Emit the final value through the async iterator before completing
            this.propertyStream._pushValue([...finalValue]);
            this.propertyStream._complete();
            this.completer.resolve(finalValue);
            this._isClosed = true;
        }
    }
}

File: packages/json/llm-json-stream-python/src/llm_json_stream/property_stream_controller.py
"""Property stream controllers for managing stream state."""

from __future__ import annotations

import asyncio
from typing import Any, Generic, List, Optional, TypeVar, TYPE_CHECKING

from .property_stream import (
    BooleanPropertyStream,
    ListPropertyStream,
    MapPropertyStream,
    NullPropertyStream,
    NumberPropertyStream,
    PropertyStream,
    StringPropertyStream,
)

if TYPE_CHECKING:  # pragma: no cover - typing only
    from .json_stream_parser import JsonStreamParserController


T = TypeVar("T")


class PropertyStreamController(Generic[T]):
    property_stream: PropertyStream[T]

    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        self._is_closed = False
        self._final_value: Optional[T] = None
        self._has_final_value = False
        self._parser_controller = parser_controller
        self._property_path = property_path
        loop = asyncio.get_event_loop()
        self._future: asyncio.Future[T] = loop.create_future()

    @property
    def is_closed(self) -> bool:
        return self._is_closed

    @property
    def has_final_value(self) -> bool:
        return self._has_final_value

    @property
    def final_value(self) -> Optional[T]:
        return self._final_value

    def complete(self, value: T) -> None:
        if self._is_closed:
            return
        self._final_value = value
        self._has_final_value = True
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(value)
        self.property_stream._notify_final_value(value)
        self._is_closed = True

    def complete_error(self, error: Exception) -> None:
        if self._is_closed:
            return
        self.property_stream._error_out(error)
        if not self._future.done():
            self._future.set_exception(error)
        self._is_closed = True


class StringPropertyStreamController(PropertyStreamController[str]):
    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        super().__init__(parser_controller, property_path)
        self._buffer = ""
        self.property_stream = StringPropertyStream(
            self._future,
            parser_controller,
            property_path,
        )

    def add_chunk(self, chunk: str) -> None:
        if self._is_closed:
            return
        self._buffer += chunk
        self.property_stream._push_value(chunk)

    def complete(self, value: str | None = None) -> None:  # type: ignore[override]
        if self._is_closed:
            return
        self._final_value = self._buffer
        self._has_final_value = True
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(self._buffer)
        self.property_stream._notify_final_value(self._buffer)
        self._is_closed = True


class NumberPropertyStreamController(PropertyStreamController[float]):
    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        super().__init__(parser_controller, property_path)
        self.property_stream = NumberPropertyStream(
            self._future,
            parser_controller,
            property_path,
        )

    def complete(self, value: float) -> None:  # type: ignore[override]
        if self._is_closed:
            return
        self._final_value = value
        self._has_final_value = True
        self.property_stream._push_value(value)
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(value)
        self.property_stream._notify_final_value(value)
        self._is_closed = True


class BooleanPropertyStreamController(PropertyStreamController[bool]):
    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        super().__init__(parser_controller, property_path)
        self.property_stream = BooleanPropertyStream(
            self._future,
            parser_controller,
            property_path,
        )

    def complete(self, value: bool) -> None:  # type: ignore[override]
        if self._is_closed:
            return
        self._final_value = value
        self._has_final_value = True
        self.property_stream._push_value(value)
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(value)
        self.property_stream._notify_final_value(value)
        self._is_closed = True


class NullPropertyStreamController(PropertyStreamController[None]):
    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        super().__init__(parser_controller, property_path)
        self.property_stream = NullPropertyStream(
            self._future,
            parser_controller,
            property_path,
        )

    def complete(self, value: None) -> None:  # type: ignore[override]
        if self._is_closed:
            return
        self._final_value = value
        self._has_final_value = True
        self.property_stream._push_value(value)
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(value)
        self.property_stream._notify_final_value(value)
        self._is_closed = True


class MapPropertyStreamController(PropertyStreamController[dict[str, Any]]):
    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        super().__init__(parser_controller, property_path)
        self._current_value: dict[str, Any] = {}
        self.property_stream = MapPropertyStream(
            self._future,
            parser_controller,
            property_path,
        )

    def notify_property(self, stream: PropertyStream[Any], key: str) -> None:
        self.property_stream._notify_property(stream, key)

    def add_snapshot(self, snapshot: dict[str, Any]) -> None:
        if self._is_closed:
            return
        self._current_value = snapshot
        self.property_stream._push_value(dict(snapshot))

    def add_property(self, key: str, value: Any) -> None:
        if self._is_closed:
            return
        self._current_value[key] = value
        self.property_stream._push_value(dict(self._current_value))

    def complete(self, value: Optional[dict[str, Any]] = None) -> None:  # type: ignore[override]
        if self._is_closed:
            return
        final_value = value if value is not None else dict(self._current_value)
        self._final_value = final_value
        self._has_final_value = True
        self.property_stream._push_value(dict(final_value))
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(final_value)
        self.property_stream._notify_final_value(final_value)
        self._is_closed = True


class ListPropertyStreamController(PropertyStreamController[list[Any]]):
    def __init__(self, parser_controller: "JsonStreamParserController", property_path: str) -> None:
        super().__init__(parser_controller, property_path)
        self._current_value: list[Any] = []
        self.property_stream = ListPropertyStream(
            self._future,
            parser_controller,
            property_path,
        )

    def notify_element(self, stream: PropertyStream[Any], index: int) -> None:
        self.property_stream._notify_element(stream, index)

    def add_snapshot(self, snapshot: list[Any]) -> None:
        if self._is_closed:
            return
        self._current_value = snapshot
        self.property_stream._push_value(list(snapshot))

    def add_element(self, value: Any) -> None:
        if self._is_closed:
            return
        self._current_value.append(value)
        self.property_stream._push_value(list(self._current_value))

    def complete(self, value: Optional[list[Any]] = None) -> None:  # type: ignore[override]
        if self._is_closed:
            return
        final_value = value if value is not None else list(self._current_value)
        self._final_value = final_value
        self._has_final_value = True
        self.property_stream._push_value(list(final_value))
        self.property_stream._complete()
        if not self._future.done():
            self._future.set_result(final_value)
        self.property_stream._notify_final_value(final_value)
        self._is_closed = True
